bpf: switch egress gateway logic to identity_is_cluster()

jibi opened the pull request:
See individual commits.

Fixes: https://github.com/cilium/cilium/issues/19784

jibi committed:
bpf: identity: add is_cluster_identity()

This function will be used by the egress gateway logic in a subsequent
commit to determine if a given identity belongs or not to an entity in
the cluster.

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi committed:
bpf: switch egress gateway logic to identity_is_cluster()

Currently to determine if a given destination IP belongs to the cluster,
the egress gateway datapath logic uses is_cluster_destination(), which
relies on different sources of truth (tunnel_endpoint, security ID and EP
map).

This commit switches that logic to the new identity_is_cluster(),
which relies solely on security IDs.

Fixes: #19784

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

jibi commented on bpf/lib/identity.h:
```diff
@@ -6,6 +6,21 @@
 #include "dbg.h"
 
+#define LOCAL_CLUSTER_IDENTITY_RANGE_START 1
+#define LOCAL_CLUSTER_IDENTITY_RANGE_END   ((1 << 16) - 1)
```
Not sure if it may be a good idea to define these in `pkg/identity`? (initially I thought I would find these ranges already defined in there, but so far I could only find them in the docs)

julianwiedmann replied:
Wonder if we should bake them into `bpf/node_config.h` (like eg. `WORLD_ID`), so tests can override the values.

pchaigno replied:
If they are not already defined on the agent's side, I would keep it as is (only defined in datapath). We can move them to `node_config.h` once we have an actual need for that.

ldelossa commented on bpf/lib/identity.h:
```diff
@@ -6,6 +6,21 @@
+#define SCOPE_LOCAL_CLUSTER 0
+#define SCOPE_MULTI_CLUSTER 1
+
+static __always_inline bool identity_in_range(__u32 identity, __u32 range_start, __u32 range_end)
```
This could be a macro expression that evaluates to a boolean. Tho, the always inline 
may make it a moot point.

pchaigno replied:
Yeah, end result will be the same and people usually prefer functions to macros.

ldelossa commented on bpf/lib/identity.h:
```diff
@@ -54,6 +69,49 @@ static __always_inline bool identity_is_reserved(__u32 identity)
+ * - identities belonging to a different cluster in case the scope is set to
+ *   SCOPE_LOCAL_CLUSTER
+ *
+ * For all other identities, including the following reserved ones:
```
```suggestion
* This function will return true for:
 * - ReservedIdentityHost
 * - ReservedIdentityUnmanaged
 * - ReservedIdentityHealth
 * - ReservedIdentityInit
 * - ReservedIdentityRemoteNode
 * - ReservedIdentityKubeAPIServer
 * - ReservedIdentityIngress
 * - All other identifies
```

ldelossa reviewed (approved):
LGTM, not sure if the CONST ranges you are questioning are in the right place. They seem OK defined where you have.

julianwiedmann commented on bpf/lib/identity.h:
```diff
@@ -6,6 +6,21 @@
+#define LOCAL_CLUSTER_IDENTITY_RANGE_END   ((1 << 16) - 1)
+#define MULTI_CLUSTER_IDENTITY_RANGE_START (1 << 16)
+#define MULTI_CLUSTER_IDENTITY_RANGE_END   ((1 << 24) - 1)
+#define CIDR_IDENTITY_RANGE_START          (1 << 24)
```
The [docs](https://docs.cilium.io/en/latest/internals/security-identities/) are not consistent there. The range map says `2^24`, but the description says:

> CIDR identities begin from 1 and end at 16777215, however since they’re shifted by 24, this makes their effective range 1 | (1 << 24) to 16777215 | (1 << 24) or from **16777217** to 33554431.

and
> Hence, the minimum value for a CIDR identity is 2^24 + 1.

jibi replied:
Seems like the range map is wrong :thinking:  I'll double check and update the docs then

jibi replied:
Updated :+1:

julianwiedmann commented on bpf/lib/identity.h:
```diff
@@ -54,6 +69,49 @@ static __always_inline bool identity_is_reserved(__u32 identity)
+			      MULTI_CLUSTER_IDENTITY_RANGE_START))
+		return false;
+
+	return true;
```
Just to note that old logic (`is_cluster_destination()`) defaults to "is outside cluster", while we now default to "is inside cluster". 

So if we missed to cover any corner case here, then it will probably flip its behaviour.

julianwiedmann reviewed (approved):
lgtm. Did you check if we have any similar code in the datapath, that would benefit from using `is_cluster_identity()`?

pchaigno commented on bpf/lib/identity.h:
```diff
@@ -6,6 +6,21 @@
+#define MULTI_CLUSTER_IDENTITY_RANGE_START (1 << 16)
+#define MULTI_CLUSTER_IDENTITY_RANGE_END   ((1 << 24) - 1)
+#define CIDR_IDENTITY_RANGE_START          (1 << 24)
+#define CIDR_IDENTITY_RANGE_END            ((1 << 24) + (1 << 16) - 1)
```
I'm guessing these are defined in the agent however, so maybe worth passing them via `node_config.h` to ensure we always use the same values in the agent and in the datapath.

jibi replied:
I couldn't find these constants in the agent, but let me double check

pchaigno replied:
We generate CIDR identities in the agent, so they must be at least partially defined.

jibi replied:
As a start, as far as the agent is concerned, CIDR identities are just "local" identities, so we need to look into `pkg/identity/cache/local.go`.

The `localIdentityCache` is initialized as follow: https://github.com/cilium/cilium/blob/e3b4464f3c2d35e1c3410406138682a2f07ab44c/pkg/identity/cache/allocator.go#L268

where `1` and `0xFFFFFF` are the min/max values.

Then, when we allocate a ~CIDR~ local identity, we do a bitwise or between the value returned by the allocator and `LocalIdentityFlag`: https://github.com/cilium/cilium/blob/e3b4464f3c2d35e1c3410406138682a2f07ab44c/pkg/identity/cache/local.go#L61

which is set to: https://github.com/cilium/cilium/blob/e3b4464f3c2d35e1c3410406138682a2f07ab44c/pkg/identity/numericidentity.go#L25

So we could define those min/max values as `LocalIdentity{Min,Max}ID` (although that would be a bit confusing as they are not the actual min and max IDs) and then pass `LocalIdentity{Min,Max}ID | LocalIdentityFlag` to the datapath as constants.

This also confirms that the range map in the docs is incorrect (may be the same case for the multicluster range)

jibi committed:
identity: add constants for min and max values for local (CIDR) identities

This commit defines a few constants in the identity package for the
minimal and maximal value IDs related to local (CIDR) identities:

* {Min,Max}AllocatorLocalIdentity are the minimal and maximal numeric
  identities that the localIdentityCache allocator can allocate for a
  local (CIDR) identity
* {Min,Max}LocalIdentity are the actual minimal and maximal numeric
  identity values for local (CIDR) identities, i.e. the ID returned by
  the allocator bitwise OR'ed with LocalIdentityFlag

Signed-off-by: Gilberto Bertin <jibi@cilium.io>

pchaigno reviewed (approved):
Sorry I took a bit to review :sweat: 

LGTM! Neat pull request by the way!

joestringer commented on pkg/identity/numericidentity.go:
```diff
@@ -24,6 +24,32 @@ const (
+	// Note that this does not represents the maximal value for a local
+	// identity, as the allocated ID will then be bitwise OR'ed with
+	// LocalIdentityFlag.
+	MaxAllocatorLocalIdentity = 0xFFFFF
```
:thinking: is this missing a portion of the range?

```suggestion
	MaxAllocatorLocalIdentity = 0xFFFFFF
```

jibi replied:
:scream:

joestringer reviewed (changes requested):
Just one minor thing that looks like a bug. Otherwise LGTM.

IIRC we discussed this a couple of weeks ago and the conclusion was that this is good enough for now, but we could look into having a dedicated map for this purpose in future. Did we ever write up anything somewhere as a handle for those ideas, even if we don't intend on implementing them in the short to medium term? Sometimes that kind of thing is useful to look back on when the topic comes up again.
